Fix HTML/MD parser producing empty MessageEntity

Rongronggg9 opened the pull request:
Close #3884

The `helpers.strip_text()` method has also been improved to be able to deal with overlapping entities.

****

It can be much more simplified if implement like that:
1. strip the text and get the left-stripped length and right-stripped length
2. iterate all entities **once**:
	1. delete an entity if the length is 0, otherwise
	2. offset its offset by the left-stripped length (if out-of-bounds, delete it, break), then
	3. delete if (offset + length) is out-of-bounds

I did not implement my fix like that since I am not sure if it is safe, so I just followed the original implementation. If the simplified implementation is considered safe, plz let me know and I will adopt it! :)

Rongronggg9 committed:
Fix HTML/MD parser producing empty MessageEntity

Lonami commented:
Thanks, feel free to implement your suggested simplification. Let me know if you want to do that in this or a separate PR.

Rongronggg9 commented:
> Thanks, feel free to implement your suggested simplification. Let me know if you want to do that in this or a separate PR.

Done. I've implemented it in this PR and added unit tests and detailed comments.

Lonami commented:
> and added unit tests and detailed comments

The "tests" in this library are pretty much a joke (they're not really ran often, or at all, and their coverage is really low), but if it helped you verify the code works correctly then it's fine.

Lonami commented on telethon/helpers.py:
```diff
@@ -64,40 +64,74 @@ def within_surrogate(text, index, *, length=None):
+                #  0 1 2 3|4 5 6 7 8 9 10
+                #   ^     ^           ^
+                #  o(1) lo(4)      o+l(1+9)
+                e.length = e.offset + e.length - left_offset
```
I personally find it clearer as

```python
e.length -= (left_offset - e.offset)
```

But this works too.

Lonami replied:
~~Also note that this might set the `length` to `0`. Imagine:~~

* ~~`offset = 0`~~
* ~~`left_offset = 4`~~
* ~~`length = 4`~~

~~So it would be preferable to move the `length 0` check to the end, I think.~~

No it won't because `offset + length` is not greater than `left_offset` so it never enters here.

Lonami commented:
Thanks!
